system/lib/libc/dynlink.c

```diff
@@ -263,7 +189,272 @@ static void ensure_init() {
   do_write_unlock();
 }
 
-void* dlopen(const char* file, int flags) {
+#ifdef _REENTRANT
+// When we are attempting the syncronize loaded libraries between threads we
+// currently abort, rather than rejecting the promises.  We could reject the
+// promises, and attempt to return an error from the original dlopen() but we
+// would have to also unwind the state on all the threads that were able to load
+// the module.
+#define ABORT_ON_SYNC_FAILURE 1
+
+// These functions are defined in JS.
+
+// Synchronous version of "dlsync_threads".  Called only on the main thread.
+// Runs _emscripten_dlsync_self on each of the threads that are running at
+// the time of the call.
+void _emscripten_dlsync_threads();
+
+// Asynchronous version of "dlsync_threads".  Called only on the main thread.
+// Runs _emscripten_dlsync_self on each of the threads that are running at
+// the time of the call.  Once this is done the callback is called with the
+// given em_proxying_ctx.
+void _emscripten_dlsync_threads_async(pthread_t calling_thread,
+                                      void (*callback)(em_proxying_ctx*),
+                                      em_proxying_ctx* ctx);
+
+static void dlsync_next(struct dlevent* dlevent, em_promise_t promise);
+
+static void sync_one_onsuccess(struct dso* dso, void* user_data) {
+  em_promise_t promise = (em_promise_t)user_data;
+  dbg("sync_one_onsuccess dso=%p event=%p promise=%p", dso, dso->event, promise);
+  // Load the next dso in the list
+  thread_local_tail = dso->event;
+  dlsync_next(thread_local_tail->next, promise);
+}
+
+static void sync_one_onerror(struct dso* dso, void* user_data) {
+#if ABORT_ON_SYNC_FAILURE
+  abort();
+#else
+  em_promise_t promise = (em_promise_t)user_data;
+  emscripten_promise_reject(promise);
+#endif
+}
+
+// Called on the main thread to asynchronously "catch up" with all the DSOs
+// that are currently loaded.
+static void dlsync_next(struct dlevent* dlevent, em_promise_t promise) {
+  dbg("dlsync_next event=%p promise=%p", dlevent, promise);
+  while (dlevent && dlevent->sym_index != -1) {
```

Comment:
Can't this list have a mixed order of dlopens and dlsyms? I'm not sure why it's ok to read all the dlsyms and then stop at the very first dlopen.

Reply:
Its not just reading the dlsyms, its synchronously processing them, until we find a dlopen event.  I've added a comment here.